### DISCUSSION TITLE
How to mark syscall eBPF programs as sleepable

### DISCUSSION THREAD

**Discussion by christian-2** [cilium/ebpf#1327] [Q&A] [answered]
I have placed an eBPF progam in an ELF section `syscall`, however when I try to load it into the kernel [`ebpf.NewCollection()`](https://pkg.go.dev/github.com/cilium/ebpf#NewCollection) raises an error: "Syscall programs can only be sleepable". From [`elfSectionDefs`](https://github.com/cilium/ebpf/blob/main/elf_sections.go#L57) I would have estimated that `syscall` already implies `_SEC_SLEEPABLE` (hence sleepable). Apparently this is not the case. How am I supposed to mark my eBPF program as sleepable.

**Comment by florianl** [comment]
Hi!
It looks like the [Kernel verifier](https://github.com/torvalds/linux/blob/54be6c6c5ae8e0d93a6c4641cb7528eb0b6ba478/kernel/bpf/verifier.c#L20654) returns this error. Did you set the `BPF_F_SLEEPABLE` flag in the [ProgramSpec](https://pkg.go.dev/github.com/cilium/ebpf#ProgramSpec)?

**Comment by christian-2** [comment]
@florianl Thanks for your quick reply. Yet, I could presumably set `BPF_F_SLEEPABLE` programmatically in Go before I load the eBPF program from an ELF file, but can this not be set in the eBPF program's source code as well? This would keep (almost) everything about that program in one place. My understanding so far is that a section name's suffix `.s` may serve this purpose, but this also did not work in my specific case so far.

**Comment by christian-2** [comment, marked as answer]
I think the puzzle is solved: the Go build pulls in `github.com/cilium/ebpf v0.12.3`, which apparently does not include a relevant [bug fix](https://github.com/cilium/ebpf/blob/38c3af9ffeeddd1d4f5fecf3c80babc7d1cd17e0/elf_sections.go#L57).
